karabo_data/geometry2.py

```diff
@@ -0,0 +1,483 @@
+
+class GridGeometryFragment:
+    ss_pixels = 64
+    fs_pixels = 128
+
+    # These coordinates are all (y, x), suitable for indexing a numpy array.
+    def __init__(self, corner_pos, ss_vec, fs_vec):
+        self.ss_vec = ss_vec
+        self.fs_vec = fs_vec
+        if fs_vec[0] == 0:
+            # Flip without transposing
+            fs_order = fs_vec[1]
+            ss_order = ss_vec[0]
+            self.transform = lambda arr: arr[..., ::ss_order, ::fs_order]
+            corner_shift = np.array([
+                min(ss_order, 0) * self.ss_pixels,
+                min(fs_order, 0) * self.fs_pixels
+            ])
+            self.pixel_dims = np.array([self.ss_pixels, self.fs_pixels])
+        else:
+            # Transpose and then flip
+            fs_order = fs_vec[0]
+            ss_order = ss_vec[1]
+            self.transform = lambda arr: arr.swapaxes(-1, -2)[..., ::fs_order, ::ss_order]
+            corner_shift = np.array([
+                min(fs_order, 0) * self.fs_pixels,
+                min(ss_order, 0) * self.ss_pixels
+            ])
+            self.pixel_dims = np.array([self.fs_pixels, self.ss_pixels])
+        self.corner_idx = corner_pos + corner_shift
+        self.opp_corner_idx = self.corner_idx + self.pixel_dims
+
+
+class AGIPD_1MGeometry:
+    """Detector layout for AGIPD-1M
+
+    The coordinates used in this class are 3D (x, y, z), and represent multiples
+    of the pixel size.
+    """
+    pixel_size = 2e-7  # 2e-7 metres == 0.2 mm
+    def __init__(self, modules):
+        self.modules = modules  # List of 16 lists of 8 fragments
+
+    @classmethod
+    def from_quad_positions(cls, quad_pos, asic_gap=2, panel_gap=29):
+        """Generate an AGIPD-1M geometry from quadrant positions.
+
+        This produces an idealised geometry, assuming all modules are perfectly
+        flat, aligned and equally spaced within their quadrant.
+
+        The quadrant positions are given in pixel units, referring to the first
+        pixel of the first module in each quadrant.
+        """
+        quads_x_orientation = [1, 1, -1, -1]
+        quads_y_orientation = [-1, -1, 1, 1]
+        modules = []
+        for p in range(16):
+            quad = p // 4
+            quad_corner = quad_pos[quad]
+            x_orient = quads_x_orientation[quad]
+            y_orient = quads_y_orientation[quad]
+            p_in_quad = p % 4
+            corner_y = quad_corner[1] - (p_in_quad * (128 + panel_gap))
+
+            tiles = []
+            modules.append(tiles)
+
+            for a in range(8):
+                corner_x = quad_corner[0] + x_orient * (64 + asic_gap) * a
+                tiles.append(AGIPDGeometryFragment(
+                    corner_pos=np.array([corner_x, corner_y, 0.]),
+                    ss_vec=np.array([x_orient, 0, 0]),
+                    fs_vec=np.array([0, y_orient, 0]),
+                ))
+        return cls(modules)
+
+    @classmethod
+    def from_crystfel_geom(cls, filename):
```

Comment:
Can this class be extended to work with other geometry format or it does not really makes sens?

if so, it would maybe be nice to have something like:
```python
from_file(filename, format='crystfel')
```

Reply:
It might make sense to support other file formats; my plan was to have a separate method for each, like `from_crystfel_geom` and `from_euxfel_geom`. There would also be corresponding methods for writing geometry files, like `write_crystfel_geom`, `write_pyfai_distortion`...

I prefer this to a `format=` parameter, because it's easy to tab-complete methods in a notebook or an IDE, and because some formats might want parameters which may not make sense for all formats.